Python/fileutils.c

```diff
@@ -1812,18 +1811,17 @@ _Py_fopen_obj(PyObject *path, const char *mode)
         Py_END_ALLOW_THREADS
     } while (f == NULL
              && errno == EINTR && !(async_err = PyErr_CheckSignals()));
-
-    Py_DECREF(bytes);
 #endif
-    if (async_err)
-        return NULL;
-
-    if (f == NULL) {
+    if (f == NULL && !async_err) {
         PyErr_SetFromErrnoWithFilenameObject(PyExc_OSError, path);
-        return NULL;
     }
+#ifdef MS_WINDOWS
+    PyMem_Free(wpath);
+#else
+    Py_DECREF(bytes);
+#endif
 
-    if (set_inheritable(fileno(f), 0, 1, NULL) < 0) {
+    if (f != NULL && set_inheritable(fileno(f), 0, 1, NULL) < 0) {
```

Comment:
I dislike changes in this file :-( I prefer to exit ASAP on error.

Reply:
Done.